R/github-pages.R

```diff
@@ -0,0 +1,164 @@
+#'   `path = "/"`. These are the usethis defaults, but more importantly, they
+#'   are GitHub's defaults, i.e. when a `gh-pages` branch is first created, it
+#'   is *automatically* published to Pages, using the source found in `"/"`. If
+#'   a `gh-pages` branch does not yet exist on the host, `use_github_pages()`
+#'   creates an empty, orphan remote branch.
+#'
+#'   The most common alternative is the repo's default branch, coupled with
+#'   `path = "/docs"`. It is the user's responsibility to ensure that this
+#'   `branch` pre-exists on the host.
+#' @param path Path to the site source. Note that GitHub only supports `"/"` or
+#'   `"/docs".`
+
+#' @param cname Optional, custom domain name. The `NA` default means "don't set
+#'   or change this", whereas a value of `NULL` removes any previously
+#'   configured custom domain.
+#'
+
+#' @seealso
+#' * <https://docs.github.com/en/free-pro-team@latest/github/working-with-github-pages>
+#' * <https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#pages>
+
+#' @return Nothing
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' use_github_pages()
+#' }
+use_github_pages <- function(branch = "gh-pages", path = "/", cname = NA) {
+  stopifnot(is_string(branch), is_string(path))
+  stopifnot(is.na(cname) || is.null(cname) || is_string(cname))
+  path <- match.arg(path, choices = c("/", "/docs"))
+  tr <- target_repo(github_get = TRUE)
+  if (!isTRUE(tr$can_push)) {
+    ui_stop("
+      You don't seem to have push access for {ui_value(tr$repo_spec)}, which \\
+      is required to turn on GitHub Pages.")
+  }
+
+  gh <- function(endpoint, ...) {
+    gh::gh(
+      endpoint,
+      ...,
+      owner = tr$repo_owner, repo = tr$repo_name, .api_url = tr$api_url
+    )
+  }
+  safe_gh <- purrr::safely(gh)
+
+  if (branch == "gh-pages") {
+    new_branch <- create_gh_pages_branch(tr, branch = "gh-pages")
+    if (new_branch) {
+      # merely creating gh-pages branch automatically activates publishing
+      # BUT we need to give the servers time to sync up before a new GET
+      # retrieves accurate info... ask me how I know
+      Sys.sleep(2)
+    }
+  }
+
+  site <- safe_gh("GET /repos/{owner}/{repo}/pages")[["result"]]
+
+  if (is.null(site)) {
+    ui_done("Activating GitHub Pages for {ui_value(tr$repo_spec)}")
+    # TODO: make sure this does actually return the site
+    site <- gh(
+      "POST /repos/{owner}/{repo}/pages",
+      source = list(branch = branch, path = path),
+      .accept = "application/vnd.github.switcheroo-preview+json"
+    )
+  }
+
+  need_update <-
+    site$source$branch != branch ||
+    site$source$path != path ||
+    (is.null(cname) && !is.null(site$cname)) ||
+    (is_string(cname) && cname != site$cname)
+
+  if (need_update) {
+    args <- list(
+      endpoint = "PUT /repos/{owner}/{repo}/pages",
+      source= list(branch = branch, path = path)
+    )
+    if (is.null(cname) && !is.null(site$cname)) {
```

Comment:
I'm not sure we should touch `cname` because it's confusing that it creates a fill (`docs/CNAME`) that will appear next time you pull.

Reply:
I lean towards keeping it. Our primary use case is with an orphan `gh-pages` branch which most of us never pull or inspect in any way.

The only properties one can set about GitHub Pages are branch, path, cname, so it just feels odd to offer a programmatic way to set/reset 2 of them but not the third. Plus we really do need this to do complete setup with `use_tidy_pkgdown()`.

Reply:
The thing I am currently mulling over is how best to align `cname` here and `cname` in `use_tidy_pkgdown()`.

Reply:
This file is usually generated by pkgdown from the `url` field and it makes me nervous to have two ways to achieving the same goal.

I don't think we need this for `use_tidy_pkgdown()`?

Reply:
So you propose leaving the `cname` arg/feature in `use_github_pages()` bug, but NOT using it when we call `use_github_pages()` inside `use_tidy_pkdown()`?

Reply:
The problem is ... someone other than us might want to use `use_tidy_pkgdown()` and ideally they'd have a way to make sure `url:` gets set in `_pkgdown.yaml`, in DESCRIPTION, in repo homepage.

Reply:
What about adding a url parameter to `use_pkgdown()`? I think we can set it up with a default, so even if you don't know the final url when you create the pkgdown site, you're at least left with a placeholder in your yaml file that you can later edit.

Reply:
I think I got to a good place here. `use_tidy_pkgdown()` no longer as any arguments, specifically, no `cname` argument.

The default pkgdown URL is now determined by what GitHub Pages tells me it is.
If the repo is owned by tidyverse or r-lib, in an interactive session, I propose that we go ahead and configure the correct, eventual URL.

All the pkgdown URL business is now also in an (unexported) helper.